pandas/core/groupby/groupby.py

```diff
@@ -1623,11 +1625,17 @@ def ohlc(self) -> DataFrame:
 
     @doc(DataFrame.describe)
     def describe(self, **kwargs):
-        with _group_selection_context(self):
-            result = self.apply(lambda x: x.describe(**kwargs))
-            if self.axis == 1:
-                return result.T
-            return result.unstack()
+        result = self.apply(lambda x: x.describe(**kwargs))
+        if self.axis == 1:
+            return result.T
+        # GH 34656 self.apply() will return non-nuisance grouping columns, but
+        # we remove them from describe if as_index=True
+        if self.as_index:
+            with _group_selection_context(self):
+                if self._group_selection is not None:
+                    cols = result.columns.intersection(self._group_selection)
+                    result = result.reindex(columns=cols)
```

Comment:
This will perhaps break for duplicates in the columns. Can you check?

```python
In [13]: df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=['A', 'B', 'B'])

In [14]: df.groupby('A', as_index=False).describe()
Out[14]:
      A                                       B
  count mean std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max count mean std  min  25%  50%  75%  max
0   1.0  1.0 NaN  1.0  1.0  1.0  1.0  1.0   1.0  2.0 NaN  2.0  2.0  2.0  2.0  2.0   1.0  3.0 NaN  3.0  3.0  3.0  3.0  3.0
1   1.0  4.0 NaN  4.0  4.0  4.0  4.0  4.0   1.0  5.0 NaN  5.0  5.0  5.0  5.0  5.0   1.0  6.0 NaN  6.0  6.0  6.0  6.0  6.0
```

Reply:
That is a good point @TomAugspurger . 

I've updated the code to handle this and added a test for it: `test_function.py::test_describe_with_duplicate_output_column_names`. None of the existing tests caught this.

Reply:
Following this update the `.describe` method is no longer affected by this PR. 

https://github.com/pandas-dev/pandas/pull/35314#issuecomment-664193224

I have kept the new test `test_function.py::test_describe_with_duplicate_output_column_names` as it makes sense to have a test for duplicate column names just to identify any future breaking changes.